internal/graph/checkdispatchset.go

```diff
@@ -0,0 +1,166 @@
+	// May be nil.
+	caveat *core.ContextualizedCaveat
+}
+
+// relationRef is a tuple of a namespace and a relation.
+type relationRef struct {
+	namespace string
+	relation  string
+}
+
+// subjectRef is a tuple of a namespace, an object ID, and a relation.
+type subjectRef struct {
+	namespace string
+	objectID  string
+	relation  string
+}
+
+// newCheckDispatchSet creates and returns a new checkDispatchSet.
+func newCheckDispatchSet() *checkDispatchSet {
+	return &checkDispatchSet{
+		bySubjectType: map[relationRef]map[string]bool{},
+		bySubject:     mapz.NewMultiMap[subjectRef, resourceIDAndCaveat](),
+	}
+}
+
+// Add adds the specified ObjectAndRelation to the set.
+func (s *checkDispatchSet) addForRelationship(tpl *core.RelationTuple) {
+	// Add an entry for the subject pointing to the resource ID and caveat for the subject.
+	riac := resourceIDAndCaveat{
+		resourceID: tpl.ResourceAndRelation.ObjectId,
+		caveat:     tpl.Caveat,
+	}
+	subjectRef := subjectRef{
+		namespace: tpl.Subject.Namespace,
+		objectID:  tpl.Subject.ObjectId,
+		relation:  tpl.Subject.Relation,
+	}
+	s.bySubject.Add(subjectRef, riac)
+
+	// Add the subject ID to the map of subjects for the type of subject.
+	siac := subjectIDAndHasCaveat{
+		objectID:           tpl.Subject.ObjectId,
+		hasIncomingCaveats: tpl.Caveat != nil && tpl.Caveat.CaveatName != "",
+	}
+	subjectTypeRef := relationRef{namespace: tpl.Subject.Namespace, relation: tpl.Subject.Relation}
+
+	subjectIDsForType, ok := s.bySubjectType[subjectTypeRef]
+	if !ok {
+		subjectIDsForType = make(map[string]bool)
+		s.bySubjectType[subjectTypeRef] = subjectIDsForType
+	}
+
+	// If a caveat exists for the subject ID in any branch, the whole branch is considered caveated.
+	subjectIDsForType[tpl.Subject.ObjectId] = siac.hasIncomingCaveats || subjectIDsForType[tpl.Subject.ObjectId]
+}
+
+func (s *checkDispatchSet) dispatchChunks(dispatchChunkSize uint16) []checkDispatchChunk {
+	// Start with an estimate of one chunk per type, plus one for the remainder.
+	expectedNumberOfChunks := len(s.bySubjectType) + 1
+	toDispatch := make([]checkDispatchChunk, 0, expectedNumberOfChunks)
+
+	// For each type of subject, create chunks of the IDs over which to dispatch.
+	for subjectType, subjectIDsAndHasCaveats := range s.bySubjectType {
+		entries := make([]subjectIDAndHasCaveat, 0, len(subjectIDsAndHasCaveats))
+		for objectID, hasIncomingCaveats := range subjectIDsAndHasCaveats {
+			entries = append(entries, subjectIDAndHasCaveat{objectID: objectID, hasIncomingCaveats: hasIncomingCaveats})
+		}
+
+		// Sort the list of subject IDs by whether they have caveats and then the ID itself.
+		sort.Slice(entries, func(i, j int) bool {
+			iHasCaveat := entries[i].hasIncomingCaveats
+			jHasCaveat := entries[j].hasIncomingCaveats
+			if iHasCaveat == jHasCaveat {
+				return entries[i].objectID < entries[j].objectID
+			}
+			return iHasCaveat && !jHasCaveat
+		})
+
+		chunkCount := 0.0
```

Comment:
The float here surprises me - is that intentional?

Reply:
Yes; we use it for the Prometheus metric below